google-api-client/src/main/java/com/google/api/client/googleapis/services/AbstractGoogleClient.java

```diff
@@ -328,9 +439,35 @@ protected Builder(
         HttpRequestInitializer httpRequestInitializer) {
       this.transport = Preconditions.checkNotNull(transport);
       this.objectParser = objectParser;
-      setRootUrl(rootUrl);
-      setServicePath(servicePath);
+      this.rootUrl = normalizeRootUrl(rootUrl);
+      this.servicePath = normalizeServicePath(servicePath);
       this.httpRequestInitializer = httpRequestInitializer;
+      this.serviceName = parseServiceName(rootUrl);
+      this.isUserConfiguredEndpoint =
```

Comment:
Why do we need to determine if the `rootUrl` is user configured based on if it ends with GDU in the `Builder`? Is setting `this.isUserConfiguredEndpoint = true;` in the setter not enough ?
Is it for the scenarios that customers may extend this class and call this Builder directly? Because I think in the public client, `rootUrl` is hidden. e.g. in [Bigquery](https://github.com/googleapis/google-api-java-client-services/blob/a3b11763afceda22d5b82b7550bf8d74617464b9/clients/google-api-services-bigquery/v2/2.0.0/com/google/api/services/bigquery/Bigquery.java#L8012-L8013)

Reply:
> Is it for the scenarios that customers may extend this class and call this Builder directly? 

Yep, it's for the (hopefully) rare user use-case where users are directly extending this class and calling the builder. The apiary libraries don't expose the endpoint configurations in the Builder directly.